rest/src/main/scala/io/udash/rest/raw/RawQueryValue.scala

```diff
@@ -0,0 +1,183 @@
+package rest.raw
+
+import com.avsystem.commons._
+import com.avsystem.commons.misc.ImplicitNotFound
+import com.avsystem.commons.rpc.{AsRaw, AsReal}
+import io.udash.utils.URLEncoder
+
+import scala.annotation.{implicitNotFound, tailrec}
+import scala.collection.AbstractIterator
+import scala.collection.generic.CanBuildFrom
+
+/**
+  * The raw data type which all query parameter values are serialized into, i.e. the RPC macro engine looks
+  * for `AsRaw/AsReal[RawQueryValue, T]` for every [[io.udash.rest.Query Query]] parameter of type `T`.
+  *
+  * Represents an URI query parameter (or sequence of adjacent parameters with the same name),
+  * in a raw form which makes it possible to support repeated, flag and binary parameters.
+  * Unlike [[PlainValue]], raw strings inside `RawQueryValue` must be already URL-encoded.
+  *
+  * Examples:
+  * `&flag` -> `Single(Opt.Empty)`
+  * `&key=value` -> `Single(Opt("value"))`
+  * `&key=value1&key=value2` -> `Multiple(Opt("value1"), Single(Opt("value2"))`
+  * `&key=va%7Clue` -> `Single(Opt("va%7Clue")`
+  */
+sealed abstract class RawQueryValue {
+
+  import RawQueryValue._
+
+  def first: Opt[String] = this match {
+    case Single(encoded) => encoded
+    case Repeated(encoded, _) => encoded
+  }
+
+  def iterator: Iterator[Opt[String]] = new AbstractIterator[Opt[String]] {
+    private var _next: RawQueryValue = RawQueryValue.this
+    def hasNext: Boolean = _next != null
+    def next(): Opt[String] = _next match {
+      case null => throw new NoSuchElementException
+      case Single(v) =>
+        _next = null
+        v
+      case Repeated(v, tail) =>
+        _next = tail
+        v
+    }
+  }
+
+  def toList: List[Opt[String]] = iterator.toList
+
+  def encodeParam(name: String): String =
+    encodeParam(name, new StringBuilder).result()
+
+  def encodeParam(name: String, sb: StringBuilder): StringBuilder = {
+    val encName = urlEncode(name)
+    def append(enc: Opt[String]): StringBuilder = enc match {
+      case Opt.Empty => sb.append(encName)
+      case Opt(v) => sb.append(encName).append("=").append(v)
+    }
+    @tailrec def loop(rqv: RawQueryValue): StringBuilder = rqv match {
+      case Single(v) => append(v)
+      case Repeated(h, t) =>
+        append(h)
+        sb.append("&")
+        loop(t)
+    }
+    loop(this)
+  }
+
+  def exploded: RawQueryValue = RawQueryValue.fromRaw {
+    val it = iterator.flatMap {
+      case Opt.Empty | Opt("") => Iterator.empty
+      case Opt(v) => v.split(",").iterator.map(Opt(_))
+    }
+    if (it.nonEmpty) it else Iterator(Opt.Empty)
+  }
+
+  def unexploded: RawQueryValue =
+    Single(Opt(iterator.flatten.mkString(",")))
+
+  def firstPlainValue: PlainValue =
+    PlainValue(first.fold("")(urlDecode))
+}
+object RawQueryValue {
+  def fromRaw(values: TraversableOnce[Opt[String]]): RawQueryValue =
+    if (values.isEmpty) throw new IllegalArgumentException("no values")
+    else values.foldRight(null: RawQueryValue) {
```

Comment:
`foldRight` here requires materializing an intermediate reversed collection. Can you enforce order at other level or ensure access to reverse iterator?